### PR TITLE
Add transformer for moving keyframes to amp-keyframes

### PR DESCRIPTION
**Pull request by ijjk**
While integrating AMP into Next.js we are looking at utilizing `optimizer` to help optimize AMP pages. I ended up integrating moving `@keyframes`, `@media`, and `@supports` to the `amp-keyframes` style tag according to the docs I found [here](https://www.ampproject.org/docs/fundamentals/spec#keyframes-stylesheet). Let me know if something needs updating, thanks!

Closes: #214

### CODE REVIEW

**Comment by ijjk** [comment]
I updated the branch to use the CSS parser, it does seem like a better idea than using regex and I added a couple of tests for it.

File: packages/optimizer/lib/transformers/SeparateKeyframes.js [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+      }
+      return !doMove;
+    });
+
+    // if no rules moved nothing to do
+    if (!keyframesTree.stylesheet.rules.length) return;
+    let hadKeyframesTag = true;
+
+    if (!stylesKeyframesTag) {
+      stylesKeyframesTag = body.firstChildByTag('style');
```
Comment (sebastianbenz) [inline packages/optimizer/lib/transformers/SeparateKeyframes.js]: Do we need to search again for the `style` tag? I thought we identified all style tags at the beginning.

File: packages/optimizer/spec/transformers/SeparateKeyframes/moves_keyframes/input.html [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+  <script async custom-element=amp-experiment src=https://cdn.ampproject.org/v0/amp-experiment-0.1.js></script>
+  <script async src="https://cdn.ampproject.org/v0.js"></script>
+  <link rel=stylesheet href=https://cdn.ampproject.org/v0.css>
+  <link href=https://fonts.googleapis.com/css?foobar rel=stylesheet type=text/css>
+  <link href=https://example.com/favicon.ico rel=icon>
+  <style amp-custom>
+    .another{display:block}@keyframes anim1{from{opacity:0}top{opacity:1}}.foo{height:0}@media screen and (max-width:960px){@keyframes anim1{from{opacity:1}to{opacity:0}}}.bar{height:auto}@supports (display:grid){div{display:grid}}
+  </style>
+</head>
+<body></body>
```
[the remark this answers is not in the document]
Comment (ijjk) [reply packages/optimizer/spec/transformers/SeparateKeyframes/moves_keyframes/input.html]: Yup, updated

File: packages/optimizer/spec/transformers/SeparateKeyframes/moves_keyframes_with_existing/input.html [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  <link rel=stylesheet href=https://cdn.ampproject.org/v0.css>
+  <link href=https://fonts.googleapis.com/css?foobar rel=stylesheet type=text/css>
+  <link href=https://example.com/favicon.ico rel=icon>
+  <style amp-custom>
+    .another{display:block}@keyframes anim1{from{opacity:0}top{opacity:1}}.foo{height:0}@media screen and (max-width:960px){@keyframes anim1{from{opacity:1}to{opacity:0}}}.bar{height:auto}@supports (display:grid){div{display:grid}}
+  </style>
+</head>
+<body>
+  <style amp-keyframes>
+    @keyframes exisiting {
```
[the remark this answers is not in the document]
Comment (ijjk) [reply packages/optimizer/spec/transformers/SeparateKeyframes/moves_keyframes_with_existing/input.html]: Oh good catch

File: packages/optimizer/spec/transformers/SeparateKeyframes/moves_keyframes_with_existing/expected_output.html [hunk 4]
```diff
@@ -0,0 +1 @@
+<!DOCTYPE html><html ⚡=""><head><script async="" custom-element="amp-experiment" src="https://cdn.ampproject.org/v0/amp-experiment-0.1.js"></script><script async="" src="https://cdn.ampproject.org/v0.js"></script><link rel="stylesheet" href="https://cdn.ampproject.org/v0.css"><link href="https://fonts.googleapis.com/css?foobar" rel="stylesheet" type="text/css"><link href="https://example.com/favicon.ico" rel="icon"><style amp-custom="">.another{display:block;}.foo{height:0;}.bar{height:auto;}</style></head><body><style amp-keyframes="">@keyframes exisiting{from{height:0px;}to{height:100px;}}@keyframes anim1{from{opacity:0;}top{opacity:1;}}@media screen and (max-width:960px){@keyframes anim1{from{opacity:1;}to{opacity:0;}}}@supports (display:grid){div{display:grid;}}</style></body></html>
```
[the remark this answers is not in the document]
Comment (ijjk) [reply packages/optimizer/spec/transformers/SeparateKeyframes/moves_keyframes_with_existing/expected_output.html]: Yeah, good point. Updated.

File: packages/optimizer/lib/transformers/SeparateKeyframes.js (hunk 1 above)
Comment (ijjk) [reply packages/optimizer/lib/transformers/SeparateKeyframes.js]: At the beginning I'm checking the head for the `style[amp-custom]` and also checking to see if there's a `style[amp-keyframes]` tag. 

Here I'm checking if a `style[amp-keyframes]` tag wasn't found in the head then I'm making sure there isn't one in the body

**Comment by ijjk** [comment]
> one test that ensures that @media and @supports are not moved

Don't we want to move `@media` and `@supports` to the `style[amp-keyframes]` also?

**Comment by sebastianbenz** [comment]
> Don't we want to move @media and @supports to the style[amp-keyframes] also?

This is where it get's tricky. If I understand the spec correctly, these may only be moved if they contain a keyframe animation (we might have to split the content of an @media / @support element). Which makes sense, as @media without keyframe are needed for the initial render of the page.

File: packages/optimizer/lib/transformers/SeparateKeyframes.js (hunk 1 above)
Comment (sebastianbenz) [reply packages/optimizer/lib/transformers/SeparateKeyframes.js]: Ah got, sorry missed that.

**Comment by ijjk** [comment]
Oh gotcha, that makes sense. I updated it to handle moving only the `@keyframes` from inside of `@supports` and `@media`

File: packages/optimizer/spec/transformers/SeparateKeyframes/moves_keyframes_with_existing/input.html [hunk 5]
```diff
@@ -0,0 +1,22 @@
+<!doctype html>
+<html ⚡>
+<head>
+  <script async custom-element=amp-experiment src=https://cdn.ampproject.org/v0/amp-experiment-0.1.js></script>
+  <script async src="https://cdn.ampproject.org/v0.js"></script>
+  <link rel=stylesheet href=https://cdn.ampproject.org/v0.css>
+  <link href=https://fonts.googleapis.com/css?foobar rel=stylesheet type=text/css>
+  <link href=https://example.com/favicon.ico rel=icon>
+  <style amp-custom>
+    .another{display:block}@keyframes anim1{from{opacity:0}top{opacity:1}}.foo{height:0}@media screen and (max-width:960px){@keyframes anim1{from{opacity:1}to{opacity:0}}}.bar{height:auto}
```
Comment (sebastianbenz) [inline packages/optimizer/spec/transformers/SeparateKeyframes/moves_keyframes_with_existing/input.html]: Can you add another example using `@support` with nested `@keyframe` for completeness sake.

**Comment by sebastianbenz** [comment]
Great! One last nit and then we're good to merge!

File: packages/optimizer/spec/transformers/SeparateKeyframes/moves_keyframes_with_existing/input.html (hunk 5 above)
Comment (ijjk) [reply packages/optimizer/spec/transformers/SeparateKeyframes/moves_keyframes_with_existing/input.html]: Yup, updated

**Comment by sebastianbenz** [comment]
```
Successfully published:
 - amp-toolbox-optimizer-express@0.6.2-beta.0+4e7a044
 - amp-toolbox-optimizer@0.5.2-beta.0+4e7a044
```

**Pull request merged**